tree.ts

```diff
@@ -0,0 +1,745 @@
+// Copyright Erik Weitnauer 2012-2015.
+
+/** This is an implementation of fully linked nary-trees. Each non-leaf node has an array
+of its children `children`, a reference to its left sibling `ls`, a reference to its right
+sibling `rs` and a reference to its parent `parent`.
+The Tree object is a collection of methods to handle tree structures, its not instanciated
+itself. Instead, each object can be a tree node.
+
+Most of the methods can accept both a single node or an array of nodes to work on.
+*/
+
+/// To get all static methods of the Tree object as instance methods on your
+/// object, you can make it inherit from the "TreeNode" class
+export class TreeNode {
+  children: TreeNode[];
+  parent: TreeNode | null;
+  ls: TreeNode | null;
+  rs: TreeNode | null;
+  id: string;
+  constructor() {
+    this.children = [];
+    this.parent = null;
+    this.ls = null;
+    this.rs = null;
+    this.id = Tree.uid();
+  }
+  stringify() {
+    return Tree.stringify(this);
+  }
+  clone(keep_ids, fields_to_clone) {
+    return Tree.clone(this, keep_ids, fields_to_clone);
+  }
+  get_mapping_to(target) {
+    return Tree.get_mapping_between(this, target);
+  }
+  get_1to1_mapping_to(target, strict = false) {
+    return Tree.get_1to1_mapping_between(this, target, strict);
+  }
```

Comment:
It may seem strange that I added the default parameter to `strict` as `false`. Before, it didn't have a default parameter. I really didn't want this PR to change any behavior. You might think that this change changes the behavior, because, in the documentation comment of `Tree.get_1to1_mapping_between()` it says
> by default strict=true

However, after talking about this specific change with Brokaw, I don't think the change I made will,  under any circumstance, change the behavior of the code. This is why:

Later, in the `Tree` class, it says
```
  static get_1to1_mapping_between(source_tree, target_tree, strict) {
    const map: { [id: string]: TreeNode[] } = {};
    if (arguments.length < 3) strict = true;
```
It then uses `strict` only in a is-it-truthy-or-falsey kind of way.

I designed an experiment to test out every possible input. Run this in your browser's console:
```
function g(t) {
  if(arguments.length < 1) t = true;
  console.log(
    `t = ${t}\t, which, when used as a condition is the same as ${!!t}. Also, arguments.length = ${arguments.length}`
  );
}
function f(y) { g(y) }
f(false);
f(true);
f(undefined);
f();
'done'
```
Then run that same code again put give the `f` function a default parameter value of `false`. In other words, replace line 7 with `function f(y = false) { g(y) }`. You will see that, as long as the `t` in `g(t)` is only used for its truthyness value, then the change of setting the default parameter to `false` doesn't change the behavior of the code in any  way.

Therefore, I conclude that my changing 
`get_1to1_mapping_to(target, strict) {` to 
`get_1to1_mapping_to(target, strict = false) {` doesn't, under any circumstance, change the code's behavior.